File: server/src/aggregate.ts
// server/src/aggregate.ts — pure aggregation layer over the codeburn session cache.
// Runtime dependencies: @codeburn/providers/device-roots.js plus @codeburn/models.js
// (cost repricing for rows cached without costUSD); SessionCache is a type-only
// import so the cache plumbing itself stays decoupled from codeburn at runtime.
import type { SessionCache } from '@codeburn/session-cache.js'
import { hostForSourcePath, type DeviceRoot } from '@codeburn/providers/device-roots.js'
import { billableOutputTokens, calculateCost } from '@codeburn/models.js'

export type UsageRow = {
  host: string; tool: string; model: string; ts: number; project: string
  tin: number; tout: number; tcacheRead: number; tcacheWrite: number; treason: number
  cost: number | null; costEstimated: boolean
}

export type Bucket = 'day' | 'hour'
/** total = 不分组的单一系列（主图「总计」视图用） */
export type Group = 'total' | 'host' | 'tool' | 'model' | 'project'
export type Metric = 'tokens' | 'cost' | 'calls'
export type SeriesResult = { times: number[]; series: Record<string, (number | null)[]>; unit: string }
export type RowFilter = { rangeMs: number | null; hosts: string[]; tools: string[]; models: string[]; projects: string[] }

const DAY = 24 * 3600 * 1000

type AnyRecord = Record<string, unknown>
const num = (v: unknown): number => (typeof v === 'number' && Number.isFinite(v) ? v : 0)

export function rowsFromCache(cache: SessionCache, devices: DeviceRoot[], localHost: string): UsageRow[] {
  const rows: UsageRow[] = []
  for (const [provider, section] of Object.entries(cache.providers ?? {})) {
    for (const [path, file] of Object.entries(section.files ?? {})) {
      const host = hostForSourcePath(path, localHost)
      const turns = ((file as AnyRecord).turns ?? []) as AnyRecord[]
      for (const turn of turns) {
        const calls = ((turn as AnyRecord).calls ?? []) as AnyRecord[]
        for (const call of calls) {
          const c = call as AnyRecord
          const usage = (c.usage ?? {}) as AnyRecord
          const ts = Date.parse(String(c.timestamp ?? ''))
          if (!Number.isFinite(ts)) continue
          const tin = num(usage.inputTokens), tout = num(usage.outputTokens)
          const tcacheRead = num(usage.cacheReadInputTokens), tcacheWrite = num(usage.cacheCreationInputTokens)
          const treason = num(usage.reasoningTokens)
          const tool = String(c.provider ?? provider)
          const model = String(c.model ?? 'unknown')
          const storedCost = typeof c.costUSD === 'number' && Number.isFinite(c.costUSD) ? c.costUSD : null
          // codeburn's cache-write path (parser.ts providerCallToCachedCall)
          // persists costUSD only for a whitelist of providers — zcode, codex,
          // workbuddy, … store `costUSD: undefined`. Mirror codeburn's own
          // cache-rehydration pricing (cachedCallToApiCall): billableOutputTokens
          // first so reasoning-in-output providers aren't double-billed, then
          // calculateCost. calculateCost returns 0 for unmapped models — that 0
          // is a valid cost in codeburn semantics, not a missing one, so it
          // flows through as-is instead of becoming null.
          const cost = storedCost ?? calculateCost(
            model, tin, billableOutputTokens(tool, tout, treason),
            tcacheWrite, tcacheRead, num(usage.webSearchRequests),
            c.speed === 'fast' ? 'fast' : 'standard',
            num(usage.cacheCreationOneHourTokens))
          rows.push({
            host, tool, model, ts,
            project: typeof c.project === 'string' && c.project ? c.project : 'unknown',
            tin, tout, tcacheRead, tcacheWrite, treason,
            cost,
            // CachedCall declares `isEstimated`; the brief sketch read
            // `costIsEstimated` — accept either spelling defensively. Rows
            // repriced from token counts (no stored costUSD) are estimates by
            // the same convention, hence `storedCost === null`.
            costEstimated: c.isEstimated === true || c.costIsEstimated === true || storedCost === null,
          })
        }
      }
    }
  }
  return rows
}

// token 总量口径与成本计费口径保持一致：reasoning-in-output 提供商（claude/codex/
// copilot，见 codeburn REASONING_INCLUDED_IN_OUTPUT）的 reasoningTokens 是
// outputTokens 的子集，再加一次会系统性抬高总量/趋势/矩阵；billableOutputTokens
// 与成本走同一个去重逻辑，两套数字不再打架。
export const rowTokens = (r: UsageRow): number =>
  r.tin + r.tcacheRead + r.tcacheWrite + billableOutputTokens(r.tool, r.tout, r.treason)

// 统一筛选：range + 维度值集合（空集合 = 不过滤）。所有 /api 聚合入口先过这一层，
// 保证「点选设备/工具」和时间范围在看板各区块语义一致。
export function filterRows(rows: UsageRow[], f: RowFilter, now: number): UsageRow[] {
  const cutoff = f.rangeMs !== null ? now - f.rangeMs : null
  const match = (list: string[], v: string): boolean => list.length === 0 || list.includes(v)
  return rows.filter(r =>
    (cutoff === null || r.ts >= cutoff) &&
    match(f.hosts, r.host) && match(f.tools, r.tool) &&
    match(f.models, r.model) && match(f.projects, r.project))
}

function groupValue(r: UsageRow, group: Group): string {
  if (group === 'total') return '总计'
  return group === 'host' ? r.host : group === 'tool' ? r.tool : group === 'model' ? r.model : r.project
}

function bucketStart(ts: number, bucket: Bucket): number {
  const d = new Date(ts)
  if (bucket === 'hour') { d.setMinutes(0, 0, 0); return d.getTime() }
  d.setHours(0, 0, 0, 0); return d.getTime()
}

export function series(rows: UsageRow[], opts: { bucket: Bucket; group: Group; rangeMs: number | null; metric: Metric; now: number; maxSeries?: number; maxBuckets?: number }): SeriesResult {
  let filtered = opts.rangeMs ? rows.filter(r => r.ts >= opts.now - opts.rangeMs!) : rows
  const unit = opts.bucket === 'day' ? DAY : 3600 * 1000
  // min/max via a single loop: `Math.min(...rows.map(...))` throws RangeError on
  // very large row counts and materializes two intermediate arrays.
  let minTs = Infinity, maxTs = -Infinity
  for (const r of filtered) {
    if (r.ts < minTs) minTs = r.ts
    if (r.ts > maxTs) maxTs = r.ts
  }
  const start = minTs !== Infinity ? bucketStart(minTs, opts.bucket) : bucketStart(opts.now, opts.bucket)
  const end = bucketStart(Math.max(opts.now, maxTs !== -Infinity ? maxTs : 0), opts.bucket)
  // Advance the grid by snapping to the next *local* bucket start instead of
  // adding a fixed epoch step: across a DST transition a local day is 23h/25h,
  // so a fixed step drifts off the bucket starts that rows are assigned to and
  // leaves rows without a grid cell. The raw walk keeps advancing by unit while
  // the snap fails to make progress — on a ≥25h day bucketStart(t + unit) can
  // snap back to the bucket just emitted, so iterating on the snapped value
  // would spin forever. The raw argument grows by unit each round and
  // bucketStart(x) > x − 36h, so the guard terminates within 2 rounds.
  let times: number[] = []
  for (let t = start; t <= end; ) {
    times.push(t)
    let raw = t + unit
    while (bucketStart(raw, opts.bucket) <= t) raw += unit
    t = bucketStart(raw, opts.bucket)
  }
  // 时间桶上限：range=all × hour 这类组合会产生巨型网格；只保留最近的
  // maxBuckets 个桶，窗口外的行直接不参与分组（避免其分组键挤占 Top-N）。
  if (opts.maxBuckets && times.length > opts.maxBuckets) {
    times = times.slice(-opts.maxBuckets)
    filtered = filtered.filter(r => bucketStart(r.ts, opts.bucket) >= times[0]!)
  }
  const index = new Map(times.map((t, i) => [t, i]))
  // 聚合容器统一用无原型对象：分组键来自数据（host/tool/model/project 名），
  // 普通字面量会被 'constructor' 等继承属性名击穿。
  const acc: Record<string, Array<{ sum: number; hasValue: boolean }>> = Object.create(null)
  const groupKeys = [...new Set(filtered.map(r => groupValue(r, opts.group)))].sort()
  for (const g of groupKeys) acc[g] = times.map(() => ({ sum: 0, hasValue: false }))
  for (const r of filtered) {
    const idx = index.get(bucketStart(r.ts, opts.bucket))
    // Defensive: a row whose bucket start is not on the grid is skipped instead
    // of crashing on a non-null assertion.
    if (idx === undefined) continue
    const cell = acc[groupValue(r, opts.group)]![idx]
    if (opts.metric === 'calls') { cell.sum += 1 } else if (opts.metric === 'cost') {
      if (r.cost !== null) { cell.sum += r.cost; cell.hasValue = true }
    } else { cell.sum += rowTokens(r) }
  }
  // Top-N：分组数超上限时按全期总量保留前 N-1 名，其余合并为「其他」，
  // 防止高基数维度（如 project）撑爆图例与响应体。
  let outKeys = groupKeys
  const mergeTarget = groupKeys.includes('其他') ? '其他*' : '其他'
  if (opts.maxSeries && groupKeys.length > opts.maxSeries) {
    const totalOf = (g: string): number =>
      acc[g]!.reduce((s, c) => s + (opts.metric === 'cost' && !c.hasValue ? 0 : c.sum), 0)
    const keep = [...groupKeys].sort((a, b) => totalOf(b) - totalOf(a) || (a < b ? -1 : 1)).slice(0, opts.maxSeries - 1)
    const rest = groupKeys.filter(g => !keep.includes(g))
    acc[mergeTarget] = times.map(() => ({ sum: 0, hasValue: false }))
    for (const g of rest) {
      const from = acc[g]!, to = acc[mergeTarget]!
      for (let i = 0; i < to.length; i++) {
        if (opts.metric === 'cost' && !from[i]!.hasValue) continue
        to[i]!.sum += from[i]!.sum
        to[i]!.hasValue = to[i]!.hasValue || from[i]!.hasValue
      }
    }
    outKeys = [...keep].sort()
    if (!keep.includes(mergeTarget)) outKeys.push(mergeTarget)
  }
  const out: Record<string, (number | null)[]> = Object.create(null)
  for (const g of outKeys) {
    out[g] = acc[g]!.map(c => (opts.metric === 'cost' ? (c.hasValue ? c.sum : null) : c.sum))
  }
  return { times, series: out, unit: opts.metric === 'cost' ? 'USD' : opts.metric === 'calls' ? 'calls' : 'tokens' }
}

export function overview(rows: UsageRow[]) {
  let totalTokens = 0, calls = 0, cacheRead = 0, freshInput = 0
  let cost = 0, costSeen = false, estimatedCost = 0
  const days = new Set<string>()
  // 无原型容器：键来自数据（host/tool 名），'constructor' 等名字不得命中继承属性
  const byHost: Record<string, { tokens: number; cost: number | null; calls: number }> = Object.create(null)
  const byTool: Record<string, { tokens: number; cost: number | null; calls: number }> = Object.create(null)
  const bump = (map: typeof byHost, key: string, r: UsageRow) => {
    const e = (map[key] ??= { tokens: 0, cost: null, calls: 0 })
    e.tokens += rowTokens(r)
    e.calls += 1
    if (r.cost !== null) { e.cost = (e.cost ?? 0) + r.cost }
  }
  for (const r of rows) {
    totalTokens += rowTokens(r); calls += 1
    cacheRead += r.tcacheRead; freshInput += r.tin
    if (r.cost !== null) {
      cost += r.cost; costSeen = true
      // 估算成本单独透出：未存储 costUSD / 标记 estimated 的行不能与实价混为一谈
      if (r.costEstimated) estimatedCost += r.cost
    }
    days.add(new Date(r.ts).toDateString())
    bump(byHost, r.host, r); bump(byTool, r.tool, r)
  }
  return {
    totalTokens, totalCost: costSeen ? cost : null, calls,
    cacheHitRate: freshInput + cacheRead > 0 ? cacheRead / (freshInput + cacheRead) : null,
    activeDays: days.size, byHost, byTool, estimatedCost,
  }
}

export function matrix(rows: UsageRow[], rowKey: Group, colKey: 'model') {
  const rowKeys = [...new Set(rows.map(r => groupValue(r, rowKey)))].sort()
  const colKeys = [...new Set(rows.map(r => r[colKey]))].sort()
  const ri = new Map(rowKeys.map((k, i) => [k, i])), ci = new Map(colKeys.map((k, i) => [k, i]))
  const values: (number | null)[][] = rowKeys.map(() => colKeys.map(() => null))
  for (const r of rows) {
    const i = ri.get(groupValue(r, rowKey))!, j = ci.get(r[colKey])!
    values[i]![j] = (values[i]![j] ?? 0) + rowTokens(r)
  }
  return { rowKeys, colKeys, values }
}

export function modelTable(rows: UsageRow[]) {
  // tokens = 每行按其所属工具口径（rowTokens）累加的总 token：同一模型可能来自
  // 多个工具（reasoning 口径不同），必须在行级去重后再合并，不能事后用单一工具重算
  const map = new Map<string, { model: string; calls: number; tokens: number; tin: number; tout: number; treason: number; tcacheRead: number; tcacheWrite: number; cost: number | null }>()
  for (const r of rows) {
    const e = map.get(r.model) ?? { model: r.model, calls: 0, tokens: 0, tin: 0, tout: 0, treason: 0, tcacheRead: 0, tcacheWrite: 0, cost: null }
    e.calls += 1; e.tokens += rowTokens(r)
    e.tin += r.tin; e.tout += r.tout; e.treason += r.treason; e.tcacheRead += r.tcacheRead; e.tcacheWrite += r.tcacheWrite
    if (r.cost !== null) e.cost = (e.cost ?? 0) + r.cost
    map.set(r.model, e)
  }
  // 排序键 = 展示口径的 token 总量，与「服务端已按 token 总量降序」一致
  return [...map.values()].sort((a, b) => b.tokens - a.tokens)
}

File: Sources/WattsonCore/Core/Aggregate.swift
// Core/Aggregate.swift — 会话缓存之上的纯聚合层。
// token 总量口径与成本计费口径保持一致：reasoning-in-output 提供商
// （claude/codex/copilot）的 reasoningTokens 是 outputTokens 的子集，
// rowTokens 与成本走同一套 billableOutputTokens 去重逻辑，两套数字不再打架。
import Foundation

public struct UsageRow: Sendable, Equatable {
    public var host: String
    public var tool: String
    public var model: String
    public var ts: Double
    public var project: String
    public var tin: Double
    public var tout: Double
    public var tcacheRead: Double
    public var tcacheWrite: Double
    public var treason: Double
    public var cost: Double?
    public var costEstimated: Bool

    public init(host: String, tool: String, model: String, ts: Double, project: String,
                tin: Double, tout: Double, tcacheRead: Double, tcacheWrite: Double,
                treason: Double, cost: Double?, costEstimated: Bool) {
        self.host = host
        self.tool = tool
        self.model = model
        self.ts = ts
        self.project = project
        self.tin = tin
        self.tout = tout
        self.tcacheRead = tcacheRead
        self.tcacheWrite = tcacheWrite
        self.treason = treason
        self.cost = cost
        self.costEstimated = costEstimated
    }
}

/// total = 不分组的单一系列（主图「总计」视图用）
public typealias Group = String
public typealias Bucket = String
public typealias Metric = String

public struct RowFilter: Sendable, Equatable {
    public var rangeMs: Double?
    public var hosts: [String]
    public var tools: [String]
    public var models: [String]
    public var projects: [String]

    public init(rangeMs: Double? = nil, hosts: [String] = [], tools: [String] = [],
                models: [String] = [], projects: [String] = []) {
        self.rangeMs = rangeMs
        self.hosts = hosts
        self.tools = tools
        self.models = models
        self.projects = projects
    }
}

public struct SeriesResult: Sendable, Equatable {
    public var times: [Double]
    public var series: [String: [Double?]]
    public var unit: String
}

let DAY: Double = 24 * 3600 * 1000

func aggNum(_ v: JSON?) -> Double {
    guard let n = v?.num, n.isFinite else { return 0 }
    return n
}

public func rowsFromCache(_ cache: SessionCache, _ devices: [DeviceRoot], _ localHost: String) -> [UsageRow] {
    var rows: [UsageRow] = []
    for (provider, section) in cache.providers {
        for (path, file) in section.files {
            let host = hostForSourcePath(path, roots: devices, localHost: localHost)
            for turn in file.turns {
                for call in turn.calls {
                    let usage = call.usage ?? CachedUsage()
                    guard let ts = parseJSDate(call.timestamp) else { continue }
                    let tin = usage.inputTokens, tout = usage.outputTokens
                    let tcacheRead = usage.cacheReadInputTokens
                    let tcacheWrite = usage.cacheCreationInputTokens
                    let treason = usage.reasoningTokens
                    let tool = call.provider ?? provider
                    let model = call.model ?? "unknown"
                    let storedCost = call.costUSD
                    // 未存 costUSD 的调用（原生解析器对未定价模型不落值）按
                    // billableOutputTokens + calculateCost 重估——与读取口径一致；
                    // calculateCost 对未收录模型返回 0，0 是合法成本，照实流转。
                    let cost = storedCost ?? calculateCost(
                        model: model, inputTokens: tin,
                        outputTokens: billableOutputTokens(provider: tool, outputTokens: tout, reasoningTokens: treason),
                        cacheCreationTokens: tcacheWrite, cacheReadTokens: tcacheRead,
                        webSearchRequests: usage.webSearchRequests,
                        speed: call.speed == "fast" ? "fast" : "standard",
                        oneHourCacheCreationTokens: usage.cacheCreationOneHourTokens)
                    rows.append(UsageRow(
                        host: host, tool: tool, model: model, ts: ts,
                        project: (call.project?.isEmpty == false) ? call.project! : "unknown",
                        tin: tin, tout: tout, tcacheRead: tcacheRead, tcacheWrite: tcacheWrite,
                        treason: treason, cost: cost,
                        // 重估行（无存储成本）按同一约定标为估算；isEstimated /
                        // costIsEstimated 两种拼写都接受
                        costEstimated: call.isEstimated == true || call.costIsEstimated == true || storedCost == nil))
                }
            }
        }
    }
    return rows
}

public func rowTokens(_ r: UsageRow) -> Double {
    r.tin + r.tcacheRead + r.tcacheWrite
        + billableOutputTokens(provider: r.tool, outputTokens: r.tout, reasoningTokens: r.treason)
}

/// 统一筛选：range + 维度值集合（空集合 = 不过滤）。所有 /api 聚合入口先过这一层，
/// 保证「点选设备/工具」和时间范围在看板各区块语义一致。
public func filterRows(_ rows: [UsageRow], _ f: RowFilter, _ now: Double) -> [UsageRow] {
    let cutoff = f.rangeMs.map { now - $0 }
    func match(_ list: [String], _ v: String) -> Bool { list.isEmpty || list.contains(v) }
    return rows.filter { r in
        (cutoff == nil || r.ts >= cutoff!) &&
        match(f.hosts, r.host) && match(f.tools, r.tool) &&
        match(f.models, r.model) && match(f.projects, r.project)
    }
}

func groupValue(_ r: UsageRow, _ group: Group) -> String {
    if group == "total" { return "总计" }
    switch group {
    case "host": return r.host
    case "tool": return r.tool
    case "model": return r.model
    default: return r.project
    }
}

struct SeriesCell {
    var sum: Double = 0
    var hasValue = false
}

public func series(_ rows: [UsageRow], bucket: Bucket, group: Group, rangeMs: Double?,
                   metric: Metric, now: Double, maxSeries: Int? = nil, maxBuckets: Int? = nil) -> SeriesResult {
    var filtered = rows
    if let rangeMs, rangeMs != 0 {
        filtered = filtered.filter { $0.ts >= now - rangeMs }
    }
    let unit = bucket == "day" ? DAY : 3600 * 1000
    // 单循环求 min/max：避免大数组上的多次遍历
    var minTs = Double.infinity, maxTs = -Double.infinity
    for r in filtered {
        if r.ts < minTs { minTs = r.ts }
        if r.ts > maxTs { maxTs = r.ts }
    }
    let start = minTs != .infinity ? bucketStart(minTs, bucket) : bucketStart(now, bucket)
    let end = bucketStart(max(now, maxTs != -.infinity ? maxTs : 0), bucket)
    // 时间网格按「对齐下一个本地桶起点」推进而不是固定 epoch 步长：跨 DST 时本地日
    // 是 23h/25h，固定步长会漂离行实际归属的桶起点。raw 每轮 +unit，直到对齐取得进展。
    var times: [Double] = []
    var t = start
    while t <= end {
        times.append(t)
        var raw = t + unit
        while bucketStart(raw, bucket) <= t { raw += unit }
        t = bucketStart(raw, bucket)
    }
    // 时间桶上限：只保留最近的 maxBuckets 个桶，窗口外的行不参与分组（避免其分组键挤占 Top-N）
    if let maxBuckets, times.count > maxBuckets {
        times = Array(times.suffix(maxBuckets))
        let floor = times[0]
        filtered = filtered.filter { bucketStart($0.ts, bucket) >= floor }
    }
    let index = Dictionary(uniqueKeysWithValues: times.enumerated().map { ($1, $0) })
    // 分组键来自数据（host/tool/model/project 名），Swift 字典天然无原型继承问题
    var groupKeys = Array(Set(filtered.map { groupValue($0, group) })).sorted()
    var acc: [String: [SeriesCell]] = [:]
    for g in groupKeys { acc[g] = times.map { _ in SeriesCell() } }
    for r in filtered {
        guard let idx = index[bucketStart(r.ts, bucket)] else { continue }
        let key = groupValue(r, group)
        if metric == "calls" {
            acc[key]![idx].sum += 1
        } else if metric == "cost" {
            if let cost = r.cost {
                acc[key]![idx].sum += cost
                acc[key]![idx].hasValue = true
            }
        } else {
            acc[key]![idx].sum += rowTokens(r)
        }
    }
    // Top-N：分组数超上限时按全期总量保留前 N-1 名，其余合并为「其他」，
    // 防止高基数维度（如 project）撑爆图例与响应体
    var outKeys = groupKeys
    let mergeTarget = groupKeys.contains("其他") ? "其他*" : "其他"
    if let maxSeries, groupKeys.count > maxSeries {
        func totalOf(_ g: String) -> Double {
            acc[g]!.reduce(0) { s, c in
                s + (metric == "cost" && !c.hasValue ? 0 : c.sum)
            }
        }
        let keep = groupKeys
            .sorted { a, b in
                let (ta, tb) = (totalOf(a), totalOf(b))
                return ta != tb ? ta > tb : a < b
            }
            .prefix(maxSeries - 1)
        let keepSet = Set(keep)
        let rest = groupKeys.filter { !keepSet.contains($0) }
        acc[mergeTarget] = times.map { _ in SeriesCell() }
        for g in rest {
            var from = acc[g]!
            var to = acc[mergeTarget]!
            for i in 0..<to.count {
                if metric == "cost" && !from[i].hasValue { continue }
                to[i].sum += from[i].sum
                to[i].hasValue = to[i].hasValue || from[i].hasValue
            }
            acc[mergeTarget] = to
        }
        outKeys = Array(keep).sorted()
        if !keepSet.contains(mergeTarget) { outKeys.append(mergeTarget) }
    }
    var out: [String: [Double?]] = [:]
    for g in outKeys {
        out[g] = acc[g]!.map { metric == "cost" ? ($0.hasValue ? $0.sum : nil) : $0.sum }
    }
    _ = groupKeys.count // groupKeys 仍被用于 Top-N 判定，避免误删
    return SeriesResult(times: times, series: out,
                        unit: metric == "cost" ? "USD" : metric == "calls" ? "calls" : "tokens")
}

public struct OverviewEntry: Sendable, Equatable {
    public var tokens: Double
    public var cost: Double?
    public var calls: Double
}

public struct OverviewResult: Sendable, Equatable {
    public var totalTokens: Double
    public var totalCost: Double?
    public var calls: Double
    public var cacheHitRate: Double?
    public var activeDays: Int
    public var byHost: [String: OverviewEntry]
    public var byTool: [String: OverviewEntry]
    public var estimatedCost: Double
}

public func overview(_ rows: [UsageRow]) -> OverviewResult {
    var totalTokens: Double = 0, calls: Double = 0, cacheRead: Double = 0, freshInput: Double = 0
    var cost: Double = 0, costSeen = false, estimatedCost: Double = 0
    var days = Set<String>()
    var byHost: [String: OverviewEntry] = [:]
    var byTool: [String: OverviewEntry] = [:]
    func bump(_ map: inout [String: OverviewEntry], _ key: String, _ r: UsageRow) {
        var e = map[key] ?? OverviewEntry(tokens: 0, cost: nil, calls: 0)
        e.tokens += rowTokens(r)
        e.calls += 1
        if let c = r.cost { e.cost = (e.cost ?? 0) + c }
        map[key] = e
    }
    for r in rows {
        totalTokens += rowTokens(r); calls += 1
        cacheRead += r.tcacheRead; freshInput += r.tin
        if let c = r.cost {
            cost += c; costSeen = true
            // 估算成本单独透出：未存储成本 / 标记 estimated 的行不能与实价混为一谈
            if r.costEstimated { estimatedCost += c }
        }
        days.insert(dayKey(r.ts))
        bump(&byHost, r.host, r); bump(&byTool, r.tool, r)
    }
    return OverviewResult(
        totalTokens: totalTokens,
        totalCost: costSeen ? cost : nil,
        calls: calls,
        cacheHitRate: freshInput + cacheRead > 0 ? cacheRead / (freshInput + cacheRead) : nil,
        activeDays: days.count,
        byHost: byHost, byTool: byTool,
        estimatedCost: estimatedCost)
}

public struct MatrixResult: Sendable, Equatable {
    public var rowKeys: [String]
    public var colKeys: [String]
    public var values: [[Double?]]
}

public func matrix(_ rows: [UsageRow], _ rowKey: Group, colKey: String = "model") -> MatrixResult {
    func col(_ r: UsageRow) -> String { colKey == "model" ? r.model : r.model }
    let rowKeys = Array(Set(rows.map { groupValue($0, rowKey) })).sorted()
    let colKeys = Array(Set(rows.map { col($0) })).sorted()
    let ri = Dictionary(uniqueKeysWithValues: rowKeys.enumerated().map { ($1, $0) })
    let ci = Dictionary(uniqueKeysWithValues: colKeys.enumerated().map { ($1, $0) })
    var values: [[Double?]] = rowKeys.map { _ in colKeys.map { _ in nil } }
    for r in rows {
        let i = ri[groupValue(r, rowKey)]!, j = ci[col(r)]!
        values[i][j] = (values[i][j] ?? 0) + rowTokens(r)
    }
    return MatrixResult(rowKeys: rowKeys, colKeys: colKeys, values: values)
}

public struct ModelTableEntry: Sendable, Equatable {
    public var model: String
    public var calls: Double
    public var tokens: Double
    public var tin: Double
    public var tout: Double
    public var treason: Double
    public var tcacheRead: Double
    public var tcacheWrite: Double
    public var cost: Double?
}

public func modelTable(_ rows: [UsageRow]) -> [ModelTableEntry] {
    // tokens = 每行按其所属工具口径（rowTokens）累加：同一模型可能来自多个工具
    //（reasoning 口径不同），必须在行级去重后再合并，不能事后用单一工具重算
    var map: [String: ModelTableEntry] = [:]
    for r in rows {
        var e = map[r.model] ?? ModelTableEntry(model: r.model, calls: 0, tokens: 0, tin: 0,
                                                tout: 0, treason: 0, tcacheRead: 0, tcacheWrite: 0, cost: nil)
        e.calls += 1
        e.tokens += rowTokens(r)
        e.tin += r.tin; e.tout += r.tout; e.treason += r.treason
        e.tcacheRead += r.tcacheRead; e.tcacheWrite += r.tcacheWrite
        if let c = r.cost { e.cost = (e.cost ?? 0) + c }
        map[r.model] = e
    }
    // 排序键 = 展示口径的 token 总量，与「服务端已按 token 总量降序」一致
    return map.values.sorted { $0.tokens > $1.tokens }
}
